### DISCUSSION TITLE
Dependency graph using continue_msg, but need limit on concurrency like function_node has

### DISCUSSION THREAD

**Discussion by cracauer** [oneapi-src/oneTBB#1406] [Design discussions]
Hi. I am using tbb to resolve a dependency graph using continue_msg. Working great, except that I need to be able to limit the number of concurrently executing node bodies like function_node can.

Is there a particular reason why continue_msg graphs don't have this capability? Did it just not come up yet, or is there a hard obstacle?

On a possibly/probably related note, I do not understand the example in https://oneapi-src.github.io/oneTBB/main/tbb_userguide/use_concurrency_limits.html and how function_node and continue_msg relate to each other in there.

**Comment by kboyarinov** [comment]
@cracauer, 

Thanks for your question. 
I don't know the exact reason why `continue_node` was not extended to support the concurrency limit.
@vossmjp, @aleksei-fedotov, do you know something in that regard?

Is there a reason why `function_node<continue_msg>` cannot be used for this purpose? 

Regarding the example you shared, I will try to explain it in more details. In that example, the `input_node` and `function_node` together are using a so-called push-pull protocol for forwarding messages (See more details [here](https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Flow_Graph_Message_Passing_Protocol.html)). 
The `input_node` generates one `big_object` each time the node body is executed and tries to forward it to the `function_node`. Since the `function_node` has the `rejecting` policy and concurrency limit equal to 3, the first three `big_object`s woud be accepted to work with. But when the `input_node` will try to broadcast the next `big_object`, the `function_node` would reject it. In that case, the `input_node` would work as a buffer of capacity 1, it will buffer the `big_object` internally and the edge between the `input_node` and `function_node` would be set in so-called pull-state. It means that when the `function_node` finish working on one of previously submitted object, it will try to get the "postponed" object from the `input_node` calling `try_get()` method on it which will return the item buffered in the `input_node`. `input_node` will continue executing the body and generating more `big_object`s, but the maximal number of object in work for `function_node` would be limited to 3.

**Comment by akukanov** [comment]
> Is there a particular reason why continue_msg graphs don't have this capability? Did it just not come up yet, or is there a hard obstacle?

`continue_node` is primarily targeted to build dependency graphs, where the node execution is triggered by completion signals of all its predecessors, and after completion it signals its successors. Such graphs are suppose to run to full completion before re-launching, because the dependencies are tracked simply by counting the number of incoming `continue_msg`s; if multiple executions of a dependency graph are running simultaneously, their messages may get mixed and trigger execution of a node prematurely.

That usage pattern assumes that a `continue_node` only runs once during execution of a graph, and so there is no need for concurrency limits.

**Reply by vossmjp** [reply]
Yes, as @akukanov states, it is because a `continue_node` body is executed whenever the internal reference count of received messages reaches the total number of known predecessors. A `continue_node` does not differentiate between messages received from predecessors.  If there are two predecessor  `continue_node` objects M and L connected by edges to a common successor `continue_node` N, N will execute when it receives two messages. If M sends two messages to N, N will execute even though L has not sent any messages to N. In theory, one can write a graph that depends on such behavior, but it is not the usual use case. More likely, we would want both M and L to complete before N starts. In those cases, it is not safe to overlap the execution of a `continue_node`-based graph. This makes concurrency limits of little practical value for `continue_node`.
